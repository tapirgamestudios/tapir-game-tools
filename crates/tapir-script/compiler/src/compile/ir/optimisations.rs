use std::{
    collections::HashMap,
    ops::{BitOr, BitOrAssign},
};

use petgraph::visit::Dfs;

use crate::{ast::SymbolId, compile::ir::TapIrFunction};

mod empty_phi;

#[derive(Clone, Copy, PartialEq, Eq)]
enum OptimisationResult {
    DidSomething,
    DidNothing,
}

impl BitOrAssign for OptimisationResult {
    fn bitor_assign(&mut self, rhs: Self) {
        *self = *self | rhs;
    }
}

impl BitOr for OptimisationResult {
    type Output = OptimisationResult;

    fn bitor(self, rhs: Self) -> Self::Output {
        if self == OptimisationResult::DidSomething || rhs == OptimisationResult::DidSomething {
            OptimisationResult::DidSomething
        } else {
            OptimisationResult::DidNothing
        }
    }
}

fn rename_all_variables(
    function: &mut TapIrFunction,
    renames: &HashMap<SymbolId, SymbolId>,
) -> OptimisationResult {
    if renames.is_empty() {
        return OptimisationResult::DidNothing;
    }

    let mut did_something = OptimisationResult::DidNothing;

    let mut dfs = Dfs::new(&*function, function.root);
    while let Some(block_id) = dfs.next(&*function) {
        let block = function.block_mut(block_id).expect("Could not find block");

        for symbol in block.sources_mut() {
            if let Some(renamed_symbol) = renames.get(symbol) {
                *symbol = *renamed_symbol;
                did_something = OptimisationResult::DidSomething;
            }
        }

        for symbol in block.targets_mut() {
            if let Some(renamed_symbol) = renames.get(symbol) {
                *symbol = *renamed_symbol;
                did_something = OptimisationResult::DidSomething;
            }
        }
    }

    did_something
}
