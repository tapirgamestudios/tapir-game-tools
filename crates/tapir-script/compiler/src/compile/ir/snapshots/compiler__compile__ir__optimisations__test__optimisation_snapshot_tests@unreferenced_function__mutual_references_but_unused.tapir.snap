---
source: crates/tapir-script/compiler/src/compile/ir/optimisations.rs
expression: output
input_file: crates/tapir-script/compiler/src/compile/ir/snapshot_tests/optimisations/unreferenced_function/mutual_references_but_unused.tapir
---
----------- before -------------
=======================
fn @toplevel() -> 

---- block 0 ----
temp.13 = int 9
storeprop temp.13, 0
return 

=======================
fn bar(x.2) -> int

---- block 0 ----
temp.14 = x.2
temp.15 = int 1
temp.16 = temp.14 - temp.15
temp.17 = baz(temp.16)
return temp.17

=======================
fn baz(x.3) -> int

---- block 0 ----
temp.18 = x.3
temp.19 = int 1
temp.20 = temp.18 - temp.19
temp.21 = bar(temp.20)
return temp.21


----------- remove_unreferenced_functions -------------
=======================
fn @toplevel() -> 

---- block 0 ----
temp.13 = int 9
storeprop temp.13, 0
return
